tractor/_trionics.py

```diff
@@ -297,68 +160,89 @@ def do_hard_kill(proc):
-            async with trio.open_nursery() as n:
+            async with trio.open_nursery():
                 for subactor, proc, portal in self._children.values():
-                    n.start_soon(do_hard_kill, proc)
+                    nursery.start_soon(do_hard_kill, proc)
 
         # mark ourselves as having (tried to have) cancelled all subactors
         self.cancelled = True
-        await self.wait()
-
-    async def __aexit__(self, etype, value, tb):
-        """Wait on all subactor's main routines to complete.
-        """
-        # XXX: this is effectively the (for now) lone
-        # cancellation/supervisor strategy (one-cancels-all)
-        # which exactly mimicks trio's behaviour
-        if etype is not None:
-            try:
-                # XXX: hypothetically an error could be raised and then
-                # a cancel signal shows up slightly after in which case
-                # the `else:` block here might not complete?
-                # For now, shield both.
-                with trio.CancelScope(shield=True):
-                    if etype in (trio.Cancelled, KeyboardInterrupt):
-                        log.warning(
-                            f"Nursery for {current_actor().uid} was "
-                            f"cancelled with {etype}")
-                    else:
-                        log.exception(
-                            f"Nursery for {current_actor().uid} "
-                            f"errored with {etype}, ")
-                    await self.cancel()
-            except trio.MultiError as merr:
-                if value not in merr.exceptions:
-                    raise trio.MultiError(merr.exceptions + [value])
-                raise
-        else:
-            log.debug(f"Waiting on subactors {self._children} to complete")
-            try:
-                await self.wait()
-            except (Exception, trio.MultiError) as err:
-                log.warning(f"Nursery cancelling due to {err}")
-                if self._children:
-                    with trio.CancelScope(shield=True):
-                        await self.cancel()
-                raise
-
-            log.debug(f"Nursery teardown complete")
+        self._join_procs.set()
 
 
 @asynccontextmanager
 async def open_nursery() -> typing.AsyncGenerator[ActorNursery, None]:
     """Create and yield a new ``ActorNursery``.
     """
+    # TODO: figure out supervisors from erlang
+
     actor = current_actor()
     if not actor:
         raise RuntimeError("No actor instance has been defined yet?")
 
-    # TODO: figure out supervisors from erlang
-    async with ActorNursery(actor) as nursery:
-        yield nursery
+    # XXX we use these nurseries because TRIP is doing all its stuff with
+    # an `@asynccontextmanager` which has an internal nursery *and* the
+    # task that opens a nursery **must also close it**.
+    errors: Dict[Tuple[str, str], Exception] = {}
+    async with trio.open_nursery() as da_nursery:
```

Comment:
shoudl probably comment on the difference between *da_* (daemon actor) and *ria_* (run in actor) nurseries..